src/io/asyncsocket.c

```diff
@@ -497,6 +497,105 @@ static void push_name_and_port(MVMThreadContext *tc, struct sockaddr_storage *na
+    MVMThreadContext *tc;
+    int               work_idx;
+} SocketInfo;
+
+static void socket_setup(MVMThreadContext *tc, uv_loop_t *loop, MVMObject *async_task, void *data) {
+    int r;
+
+    /* Add to work in progress. */
+    MVMObject    *arr = MVM_repr_alloc_init(tc, tc->instance->boot_types.BOOTArray);
+    MVMAsyncTask *t   = (MVMAsyncTask *)async_task;
+    SocketInfo   *si  = (SocketInfo *)data;
+    si->tc            = tc;
+    si->work_idx      = MVM_io_eventloop_add_active_work(tc, async_task);
+
+    /* Create and initialize socket. */
+    si->socket       = MVM_malloc(sizeof(uv_tcp_t));
+    si->socket->data = data;
+    if ((r = uv_tcp_init(loop, si->socket)) < 0) {
+        MVMROOT2(tc, arr, t, {
+            MVM_repr_push_o(tc, arr, t->body.schedulee);
+            MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTIO);
+
+            MVMString *msg_str = MVM_string_ascii_decode_nt(tc,
+                tc->instance->VMString, uv_strerror(r));
+            MVMObject *msg_box = MVM_repr_box_str(tc,
+                tc->instance->boot_types.BOOTStr, msg_str);
+            MVM_repr_push_o(tc, arr, msg_box);
+
+            MVM_repr_push_o(tc, t->body.queue, arr);
+        });
+
+        /* Cleanup handles. */
+        uv_close((uv_handle_t *)si->socket, free_on_close_cb);
+        si->socket = NULL;
+        MVM_io_eventloop_remove_active_work(tc, &si->work_idx);
+    } else {
+        MVMROOT2(tc, arr, t, {
+            MVM_repr_push_o(tc, arr, t->body.schedulee);
+
+            MVMOSHandle          *result = (MVMOSHandle *)MVM_repr_alloc_init(tc, tc->instance->boot_types.BOOTIO);
+            MVMIOAsyncSocketData *data   = MVM_calloc(1, sizeof(MVMIOAsyncSocketData));
+            data->handle                 = (uv_stream_t *)si->socket;
+            result->body.ops             = &op_table;
+            result->body.data            = data;
+            MVM_repr_push_o(tc, arr, (MVMObject *)result);
+            MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTStr);
+
+            MVM_repr_push_o(tc, t->body.queue, arr);
+        });
+    }
+
+}
+
+static void socket_gc_free(MVMThreadContext *tc, MVMObject *t, void *data) {
+    if (data) {
+        SocketInfo *si = (SocketInfo *)data;
+        MVM_free(si);
+    }
+}
+
+static const MVMAsyncTaskOps socket_op_table = {
+    socket_setup,
+    NULL,
+    NULL,
+    NULL,
+    socket_gc_free
+};
+
+MVMObject * MVM_io_socket_async(MVMThreadContext *tc, MVMObject *queue,
+                                       MVMObject *schedulee, MVMObject *async_type) {
+    MVMAsyncTask *task;
+    SocketInfo   *si;
+
+    /* Validate REPRS. */
+    if (REPR(queue)->ID != MVM_REPR_ID_ConcBlockingQueue)
+        MVM_exception_throw_adhoc(tc,
+            "asyncsocket target queue must have ConcBlockingQueue REPR");
+    if (REPR(async_type)->ID != MVM_REPR_ID_MVMAsyncTask)
+        MVM_exception_throw_adhoc(tc,
+            "asyncsocket result type must have REPR AsyncTask");
```

Comment:
Nitpick, but the first error message is '... type REPR' and the second is '... REPR type'.